common/txmgr/tracker.go

```diff
@@ -0,0 +1,336 @@
+	if err != nil {
+		return fmt.Errorf("failed to get enabled addresses for chain: %w", err)
+	}
+
+	if len(enabledAddrs) == 0 {
+		tr.lggr.Warnf("enabled address list is empty")
+	}
+
+	for _, addr := range enabledAddrs {
+		tr.enabledAddrs[addr] = true
+	}
+	return nil
+}
+
+// trackAbandonedTxes called once to find and insert all abandoned txes into the tracker.
+func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) trackAbandonedTxes(ctx context.Context) (err error) {
+	if tr.isStarted {
+		return fmt.Errorf("tracker already started")
+	}
+
+	nonFatalTxes, err := tr.txStore.GetNonFatalTransactions(ctx, tr.chainID)
+	if err != nil {
+		return fmt.Errorf("failed to get non fatal txes from txStore: %w", err)
+	}
+
+	// insert abandoned txes
+	for _, tx := range nonFatalTxes {
+		if !tr.enabledAddrs[tx.FromAddress] {
+			tr.insertTx(tx)
+		}
+	}
+
+	if err := tr.handleTxesByState(ctx, 0); err != nil {
+		return fmt.Errorf("failed to handle txes by state: %w", err)
+	}
+
+	return nil
+}
+
+func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) HandleTxesByState(ctx context.Context, blockHeight int64) error {
+	tr.lock.Lock()
+	defer tr.lock.Unlock()
+	tr.ctx, tr.ctxCancel = context.WithTimeout(ctx, handleTxesTimeout)
+	defer tr.ctxCancel()
+	return tr.handleTxesByState(ctx, blockHeight)
+}
+
+func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) handleTxesByState(ctx context.Context, blockHeight int64) error {
+	for id, atx := range tr.txCache {
+		tx, err := tr.txStore.GetTxByID(ctx, atx.id)
+		if err != nil {
+			return fmt.Errorf("failed to get tx by ID: %w", err)
+		}
+
+		switch tx.State {
+		case TxConfirmed:
+			if err := tr.handleConfirmedTx(tx, blockHeight); err != nil {
+				return fmt.Errorf("failed to handle confirmed txes: %w", err)
+			}
+		case TxConfirmedMissingReceipt, TxUnconfirmed:
+			// Keep tracking tx
+		case TxInProgress, TxUnstarted:
+			// Tx could never be sent on chain even once. That means that we need to sign
+			// an attempt to even broadcast this Tx to the chain. Since the fromAddress
+			// is deleted, we can't sign it.
+			errMsg := "The FromAddress for this Tx was deleted before this Tx could be broadcast to the chain."
+			if err := tr.markTxFatal(ctx, tx, errMsg); err != nil {
+				return fmt.Errorf("failed to mark tx as fatal: %w", err)
+			}
+			delete(tr.txCache, id)
+		case TxFatalError:
+			delete(tr.txCache, id)
+		default:
+			tr.lggr.Errorw(fmt.Sprintf("unhandled transaction state: %v", tx.State))
+		}
+	}
+
+	return nil
+}
+
+// handleConfirmedTx removes a transaction from the tracker if it's been finalized on chain
+func (tr *Tracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) handleConfirmedTx(
+	tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE],
+	blockHeight int64,
+) error {
+	finalized, err := tr.txStore.IsTxFinalized(context.Background(), blockHeight, tx.ID, tr.chainID)
```

Comment:
I think you can maybe leverage the ctx passed to `handleTxesByState` here so you don't have to use `context.Background()` here